src/MongoDB.Driver.Core/Core/Compression/ZlibCompressor.cs

```diff
@@ -0,0 +1,94 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.IO;
+using System.IO.Compression;
+
+namespace MongoDB.Driver.Core.Compression
+{
+    /// <summary>
+    /// Compressor according to the zlib algorithm.
+    /// </summary> 
+    internal sealed class ZlibCompressor : ICompressor
+    {
+        private readonly CompressionLevel _compressionLevel;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ZlibCompressor" /> class.
+        /// </summary>
+        /// <param name="compressionLevel">The compression level.</param>
+        public ZlibCompressor(int? compressionLevel = 0)
+        {
+            _compressionLevel = GetCompressionLevel(compressionLevel);
+        }
+
+        public bool IsAvailable => true;
+
+        /// <inheritdoc />
+        public CompressorType Type => CompressorType.Zlib;
+
+        /// <inheritdoc />
+        public byte[] Compress(byte[] bytesToCompress, int offset)
+        {
+            using (var memoryStream = new MemoryStream())
+            {
+                using (var zlibStream = new DeflateStream(memoryStream, _compressionLevel))
+                {
+                    zlibStream.Write(bytesToCompress, offset, bytesToCompress.Length - offset);
+                }
+
+                return memoryStream.ToArray();
+            }
+        }
+
+        /// <inheritdoc />
+        public byte[] Decompress(byte[] bytesToDecompress)
+        {
+            using (var memoryStream = new MemoryStream(bytesToDecompress))
+            {
+                using (var zlibStream = new DeflateStream(memoryStream, CompressionMode.Decompress))
+                using (var result = new MemoryStream())
+                {
+                    zlibStream.CopyTo(result);
+                    return result.ToArray();
+                }
+            }
+        }
+
+        private static CompressionLevel GetCompressionLevel(int? compressionLevel)
+        {
+            if (!compressionLevel.HasValue)
```

Comment:
Mapping the `spec` options to the options which are available in the .net built-in zlib compression stream.

Reply:
Would a switch statement or a Dictionary be more readable?

Reply:
I will change on `switch`.